internal/cmd/dockerbuild/main.go

```diff
@@ -1,55 +1,166 @@
 	"log"
 	"os"
+	"path/filepath"
+	"runtime"
+	"slices"
 	"strings"
 	"time"
 
 	"github.com/bufbuild/buf/private/pkg/interrupt"
-
 	"github.com/bufbuild/plugins/internal/docker"
+	"golang.org/x/mod/semver"
+	"golang.org/x/sync/errgroup"
+
 	"github.com/bufbuild/plugins/internal/plugin"
 )
 
-// dockerbuild is a helper program used to build plugins from Dockerfiles.
+// dockerbuild is a helper program used to build plugins from Dockerfiles in an optimized fashion.
+// It replaces some clunky (and non-parallel) jobs in the Makefile.
+// It knows about relationships between common containers (like protoc/grpc Bazel plugins),
+// which enables optimized builds which build common code first before binaries.
 // It also makes it easier to add new labels to images using existing code to parse buf.plugin.yaml.
 
+const (
+	// larger amount of parallelism lead to OOM errors in testing - clamp for now.
+	maxLimit = 8
+	// plugin group for all bazel builds (they are run serially on their own as they are very resource intensive).
+	bazelPluginGroup = "bazel"
+)
+
 func main() {
 	var (
 		dir      = flag.String("dir", ".", "Directory path to plugins")
 		org      = flag.String("org", "bufbuild", "Docker Organization")
 		cacheDir = flag.String("cache-dir", "", "Cache directory")
 	)
 	flag.Parse()
-	if err := run(*dir, *org, *cacheDir, flag.Args()); err != nil {
+	cmd := &command{
+		pluginsDir:      *dir,
+		dockerOrg:       *org,
+		cacheDir:        *cacheDir,
+		dockerBuildArgs: flag.Args(),
+	}
+	if err := cmd.run(); err != nil {
 		log.Fatalf("failed to build: %v", err)
 	}
 }
 
-func run(basedir string, dockerOrg string, cacheDir string, args []string) error {
+type command struct {
+	// pluginsDir specifies the directory where plugins are found (typically the root of the bufbuild/plugins repo).
+	pluginsDir string
+	// dockerOrg is the Docker organization to use in the tagged image.
+	dockerOrg string
+	// cacheDir is an optional setting to the root where Docker buildx local caches should be kept.
+	cacheDir string
+	// dockerBuildArgs contains additional arguments to pass to the Docker build.
+	dockerBuildArgs []string
+}
+
+func (c *command) run() error {
 	// Catch ctrl+c to kill the build process
 	ctx, cancel := interrupt.WithCancel(context.Background())
 	defer cancel()
-	plugins, err := plugin.FindAll(basedir)
+	allPlugins, err := plugin.FindAll(c.pluginsDir)
 	if err != nil {
 		return err
 	}
-	includedPlugins, err := plugin.FilterByPluginsEnv(plugins, os.Getenv("PLUGINS"))
+	includedPlugins, err := plugin.FilterByPluginsEnv(allPlugins, os.Getenv("PLUGINS"))
 	if err != nil {
 		return err
 	}
 	if len(includedPlugins) == 0 {
 		return nil // nothing to build
 	}
-	for _, pluginToBuild := range includedPlugins {
+	pluginGroups, err := getPluginGroups(includedPlugins)
+	if err != nil {
+		return err
+	}
+	// Build bazel plugins on their own as they are very resource intensive.
+	if bazelPlugins, ok := pluginGroups[bazelPluginGroup]; ok {
```

Comment:
Shuffled things around a bit so we build any bazel plugins first (serially). Then we build all the remaining ones in parallel up to the parallelism limit.